external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+        dx_file.close()
+
+        return 'dx://' + dx_file.id
+
+    def localize_inputs(obj, match_func, localize_func):
+        """Recursively walk through the input dict, localizing inputs as
+        necessary.
+        """
+
+        if isinstance(obj, (list, tuple)):
+            return [localize_inputs(e, match_func, localize_func) for e in obj]
+        if isinstance(obj, dict):
+            return {k: localize_inputs(v, match_func, localize_func) for k, v in obj.items()}
+        if isinstance(obj, basestring):
+            if match_func(obj):
+                return localize_func(obj)
+        return obj
+
+    localized_inputs_dict = localize_inputs(inputs_dict, match_url, localize_url)
+
+    return localized_inputs_dict
+
+
+def dx_to_wes_state(dx_state):
+    """Convert the state returned by DNAnexus to something from the list of states
+    defined by WES.
+    """
+
+    if dx_state == "in_progress":
+        return "RUNNING"
+    elif dx_state in ("partially_failed", "failed"):
+        return "EXECUTOR_ERROR"
+    elif dx_state in "done":
+        return "COMPLETE"
+
+    return "UNKNOWN"
+
+
+##############################################################################
+## Lambda functions
+##############################################################################
+
+def run_dxwdl(event, context):
+    """Run dxWDL to create a DNAnexus workflow.
+
+    This function is invoked by the function that handles POST to /workflows.
+    We'd like to expose that function via API Gateway, but API Gateway has a
+    29 second limit on a response, and dxWDL and input localization may not
+    complete in that time. But, it's pretty likely to complete in the five
+    minute lambda time limit. So, we delegate all the actual work here and
+    let POST /workflows return right away.
+
+    The inputs are all defined in the events dict:
+
+    workflow_descriptor: the main WDL string
+    workflow_params: the inputs dict for the workflow
+    workflow_dependencies: base64-encoded string of the dependencies zip
+        archive
+    auth_header: authorization header with the DNAnexus bearer token
+    dx_project_id: the DNAnexus project in which to build and run the
+        workflow
+    wes_workflow_id: the WES workflow id that was generated by the
+        workflows post function
+    """
+
+    # Is this a retry? Stop right away if so, there's no point to retrying
+    # these.
+    try:
+        if last_request_id == context.aws_request_id:
+            return
+    except NameError:
+        last_request_id = context.aws_request_id
+
+    # Clean up the /tmp directory in case we're reusing a container
+    for obj in os.listdir("/tmp"):
+        try:
+            path = os.path.join("/tmp", obj)
+            print "Deleting", path
+            if os.path.isdir(path):
+                shutil.rmtree(os.path.join("/tmp", obj))
+            else:
+                os.remove(path)
+        except OSError as e:
+            print e
```

Comment:
Is it worth logging this instead of printing it? If printed, is it recoverable somewhere?

Reply:
Got it, so it's: 
```
try:
    if dir: remove directory + all files
    else: assume file, and remove
except:
    wasn't a file, print and continue
```